From a9bded182fe6532c06fe6b40b81c8f6616757735 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Edwin=20T=C3=B6r=C3=B6k?= <edwin.torok@cloud.com>
Date: Fri, 11 Apr 2025 10:35:38 +0100
Subject: [PATCH] CA-409488: prevent Xenctrl exceptions from escaping on VM
 boot/shutdown races
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Note that this hostload metric is completely wrong, and we should
probably drop it.
But for now prevent it from introducing gaps in the entire host CPU
graph: if an exception escapes here, then nothing else gets reported,
because the host CPU, VM CPU and hostload metrics all get returned as a single
list. An exception escaping from any prevents all the others from
working too.

Fixes: 9aa7dfcca ("CP-43574: Add host load data source")
Signed-off-by: Edwin Török <edwin.torok@cloud.com>
---
 ocaml/xcp-rrdd/bin/rrdp-cpu/rrdp_cpu.ml | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/ocaml/xcp-rrdd/bin/rrdp-cpu/rrdp_cpu.ml b/ocaml/xcp-rrdd/bin/rrdp-cpu/rrdp_cpu.ml
index 7a0db5ec5..998ab7776 100644
--- a/ocaml/xcp-rrdd/bin/rrdp-cpu/rrdp_cpu.ml
+++ b/ocaml/xcp-rrdd/bin/rrdp-cpu/rrdp_cpu.ml
@@ -188,12 +188,15 @@ let dss_hostload xc domains =
   let load =
     List.fold_left
       (fun acc (dom, _, domid) ->
-        sum 0 dom.Xenctrl.max_vcpu_id (fun id ->
-            let vcpuinfo = Xenctrl.domain_get_vcpuinfo xc domid id in
-            if vcpuinfo.Xenctrl.online && not vcpuinfo.Xenctrl.blocked then
-              1
-            else
-              0
+        ( try
+            sum 0 dom.Xenctrl.max_vcpu_id (fun id ->
+                let vcpuinfo = Xenctrl.domain_get_vcpuinfo xc domid id in
+                if vcpuinfo.Xenctrl.online && not vcpuinfo.Xenctrl.blocked then
+                  1
+                else
+                  0
+            )
+          with _ -> 0
         )
         + acc
       )
